syntax = "proto3";

package example;

// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {}
  
  // Streams data back to the client
  rpc StreamData (DataRequest) returns (stream DataResponse) {}
  
  // Client streaming
  rpc ClientStream (stream ClientMessage) returns (ServerResponse) {}
  
  // Bidirectional streaming
  rpc BidirectionalStream (stream ClientMessage) returns (stream ServerResponse) {}
}

// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
  int32 status = 2;
}

// Request for streaming data
message DataRequest {
  int32 count = 1;
  string prefix = 2;
}

// Response for streaming data
message DataResponse {
  int32 id = 1;
  string value = 2;
  string timestamp = 3;
}

// Client message for streaming
message ClientMessage {
  string client_id = 1;
  string message = 2;
  int32 sequence = 3;
}

// Server response for streaming
message ServerResponse {
  string server_id = 1;
  string response = 2;
  int32 received_sequence = 3;
  string timestamp = 4;
}
